testplan/testing/base.py

```diff
@@ -366,19 +389,95 @@ def _record_teardown_start(self):
     def _record_teardown_end(self):
         self.report.timer.end("teardown")
 
-    def pre_resource_steps(self):
+    def _init_context(self):
+        if callable(self.cfg.initial_context):
+            self.resources._initial_context = self.cfg.initial_context()
+        else:
+            self.resources._initial_context = self.cfg.initial_context
+
+    def _build_environment(self, delay: bool):
+        """
+        This method will execute at most twice, once with delay=True during __init__,
```

Comment:
I do not like this two phased approach, where we prepare the function that it will be run twice. Why do we need to do the driver line up in init in case we do not have a callable?

I would turn the list during params validation into a callable and just handle the callable case in teh business logic.

Reply:
1) the 2 phase approach... is because I don't want to lose any existing test coverage for interactive mode that asserts environment status before test starts

2) I know this turn the list into callable approach, just i would prefer not to alter user input, but save it as-is in self.cfg.xxx

Reply:
I can buy your second point. I do not like the excuse of leaving complication in code to not break existing tests. Would it be to tedious to update those tests?